torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp

```
@@ -0,0 +1,1261 @@
+
+  for (auto it = writeHistory.begin(); it != writeHistory.end();) {
+    auto& indexBounds = it->first;
+    std::shared_ptr<AccessInfo> other = it->second;
+    if (info->hasDependency(other)) {
+      ++it;
+      continue;
+    }
+
+    OverlapKind overlap = overlaps(indexBounds, info->bounds());
+
+    if (overlap == NoOverlap) {
+      ++it;
+      continue;
+    }
+
+    // Only writes can close open accesses.
+    if (!isWrite) {
+      info->addDependency(other);
+      other->addDependent(info);
+      ++it;
+      continue;
+    }
+
+    // If we're not closing accesses we can stop here.
+    if ((int)other->id() > latestAccessToClose) {
+      ++it;
+      continue;
+    }
+
+    if (overlap == TotalOverlapB) {
+      // Total overlap is easy - the new access totally replaces the old.
+      it = writeHistory.erase(it);
+    } else {
+      // The new write partially overlaps a previous write. We want to keep
+      // both, but only track the unconvered part of the earlier write.
+
+      // Determine the slices of the earlier bound not covered by info.
+      auto newBounds =
+          subtractIndicesBounds(indexBounds, info->bounds(), overlap);
+
+      // Erase the old slice.
+      it = writeHistory.erase(it);
+
+      // Add all new slices.
+      for (auto& b : newBounds) {
+        it = writeHistory.insert(it, std::make_pair(b, other));
+      }
+      it++;
+    }
+  }
+
+  if (insert && isWrite) {
+    writeHistory.emplace_back(std::make_pair(info->bounds(), info));
+  }
+}
+
+void MemDependencyChecker::mergeScope(
+    const std::shared_ptr<Scope>& child,
+    const std::shared_ptr<Scope>& parent,
+    bool closeOverlapped) {
+  if (child->accesses_.empty()) {
+    return;
+  }
+
+  int latestAccessToClose =
+      closeOverlapped ? (int)child->accesses_.front()->id() : -1;
+
+  // Update dependencies, but don't add new open writes yet.
+  for (auto& info : child->accesses_) {
+    auto& writeHistory = parent->openWrites_[info->var()];
```

As in other places, using the indexing operator here makes me nervous. I'd rather use `find` and assert it's found, because if there's a bug in tracking open writes this will fail silently with the indexing operator. Or do you actually mean to open writes as a side-effect?

---

I do mean to open the container as a side effect. This is essentially a defaultdict from var->deque. I'll add a comment here.

---

If we want the side effect everywhere, comments will do for future readers as well, I think.

---

BTW I think more concise is fine too, along the lines of "NB: we want the indexing operator side-effects."